src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,164 @@
+        int oldCapacity = elementData.length;
+        if (oldCapacity * INCREMENT_VALUE > minCapacity) {
+            currentCapacity = (int) (oldCapacity * INCREMENT_VALUE);
+            return elementData = Arrays.copyOf(elementData, currentCapacity);
+        } else {
+            return elementData = Arrays.copyOf(elementData, (int) (minCapacity * INCREMENT_VALUE));
+        }
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is incorrect " + index);
+        }
+    }
+
+    public Object[] toArray(List<T> list) {
+        if (list != null) {
+            Object[] array = new Object[list.size()];
+            for (int i = 0; i < array.length; i++) {
+                array[i] = list.get(i);
+            }
+            return array;
+        }
+        throw new RuntimeException("Given object cannot be converted to an array");
+    }
+
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        final int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                s - index);
+        elementData[index] = value;
+        size = s + 1;
+    }
+
+    @Override
+    public void add(T value) {
+        add(value, elementData, size);
+    }
 
+    private void add(T value, Object[] elementData, int s) {
+        if (s == elementData.length) {
+            elementData = grow();
+        }
+        elementData[s] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] array = toArray(list);
+        int numNew = array.length;
+        Object[] elementData;
+        final int s;
+        if (numNew > (elementData = this.elementData).length - (s = size)) {
+            elementData = grow(s + numNew);
+        }
+        System.arraycopy(array, 0, elementData, s, numNew);
+        size = s + numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is incorrect " + index);
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is incorrect " + index);
+        }
+        Objects.checkIndex(index, size);
```

Comment:
The use of `Objects.checkIndex` is redundant here as you are already throwing a custom `ArrayListIndexOutOfBoundsException`. Consider removing this line to avoid unnecessary checks.